### ISSUE TITLE
Porting codebase utilizing tf.slim to TF-2.0

### ISSUE THREAD

**Issue by calledbymountains** [tensorflow/tensorflow#25227]
A lot of my TensorFlow  codebase depends upon the use of pretrained models as a part of [tensorflow/models/research/slim/nets](https://github.com/tensorflow/models/tree/master/research/slim/nets) repository.

I routinely utilize the dictionary of `end points` provided by `tf.slim` models in my codebase. Furthermore, the model definitions in the aforementioned repository is impressive and general in the sense that for models like `resnet_v2`, it allows me to directly specify the `output_stride` which leads to an automatic specification of dilation rates. 

As I aim to write my new codes in TF-2.0 ( as well as thinking of porting my existing codebase to TF-2.0), I realize (based on documentation and my knowledge of Keras !!), that there is no concept of `end point` dictionaries and `output_stride` (wherever applicable). 

Hence, I am confused about how to handle these situations. 

I believe that it could be very reasonable for the TensorFlow team to provide documentation on this issue for it could really make transition to TF-2.0,  very simple and practical.

**Comment by ymodak** [comment]
To simplify the migration to TensorFlow 2.0, there is a [conversion tool](https://github.com/tensorflow/docs/blob/master/site/en/r2/guide/upgrade.md) which updates TensorFlow 1.x Python code to use TensorFlow 2.0. Please take a look if haven't already. Thanks!

**Comment by calledbymountains** [comment]
Thank you. But does it specifically address the case of `end_point` dictionaries and `output_stride` ?

**Comment by ymodak** [comment]
@martinwicke Can you PTAL? Thanks!

**Comment by martinwicke** [comment]
It does not. In fact, it does not handle anything specific to slim. 

You can continue to use slim in 2.0, and if you are so dependent on its specifics, you may want to. slim makes heavy use of variable scopes and collections, and writing code using it will never truly feel like TensorFlow 2.0, but using tf.compat.v1 you will be able to make it work.

We will be publishing more detailed conversion guides. I will close this issue since we are already working on that (and we'll only forget to close it otherwise).

**Comment by DenisPolygalov** [comment]
@martinwicke To make it clear: 
1. The upgrade script does not handle this problem. 
2. Using tf.compat.v1 will not make it work. 
3. The only “more detailed conversion guide” I was able to find 9 month since this issue was opened is only short paragraph [here](https://www.tensorflow.org/guide/migrate) which basically state “it is not simple and it is your problem”.

Your statement was addressed to the person how opened this issue, but let me inform you how many people in fact this issue affects. If anyone at Google cares of course…

I came across this problem while trying to understand what’s going wrong with [DeepLabCut](https://github.com/AlexEMG/DeepLabCut) ( hello @AlexEMG ) software. The DeepLabCut is a marker-less pose estimation software for analyzing behavior of various subjects in a video streams. It has a [Nature paper](https://www.nature.com/articles/s41593-018-0209-y/metrics) and become so popular that it triggered “revolution in analyzing of behavior in neuroscience”. So as result, many Labs want to use DeepLabCut now because it, well, “have a Nature paper”!

Looking inside of DeepLabCut’s code unveil several severe technical problems however. These problems (so far) do not affect the main effect (e.g. pose detection) but anyone familiar with long term software support understands where it all goes…

DeepLabCut has been built on top of [Human Pose Estimation with TensorFlow](https://github.com/eldar/pose-tensorflow) project also presented in [DeeperCut](https://arxiv.org/abs/1605.03170) and [ArtTrack](https://arxiv.org/abs/1612.01465) papers. The DeeperCut rely on ResNet which was in tf.contrib.slim (TF 1.x) and therefore not available in TF 2.0 anymore...

So what do we have here. Google wrote TensorFlow 1.x and despite disastrous code quality it become popular (because it well… “Made by Google”and it is AI!). Then after TF 1.x become popular more and more people began to have severe problems (e.g. computational graph debugging issues) caused by initially disastrous design decisions. Then someone at Google finally realize that it make sense to try to fix everything, so as result TF 2.x was spitted out. As a consequence of that all code from tf.contrib was wiped out of TF 2.x betraying another Google's own creature [TF-Slim library](https://ai.googleblog.com/2016/08/tf-slim-high-level-library-to-define.html) because “it will never truly feel like TensorFlow 2.0” leaving many people dealing with many issues by themselves...

This remind me worst practice used many times by Microsoft with their various “cool APIs” dropped 2 years after rolling out.

**Comment by martinwicke** [comment]
@pichuan @tomerk How far are we with the slim port?

**Comment by pichuan** [comment]
Hi @martinwicke , I saw your mention here just now.
Short answer: I do not have an ETA.

Long answer:
After my last update here: https://github.com/google-research/tf-slim/pull/1#issuecomment-570648906 , I did spend some time trying to incorporate @tomerk 's suggestion. 

This is taking more time than I expected, and I don't think I fully understand the issue (or the solution) to give an ETA.
Note that even if/when I resolve this, there will still be 2 other pending tasks that I don't plan to take on.
Given my schedule in the next few weeks, I won't be able to look further into this right now until at least February.

**Comment by ibarrond** [comment]
While fighting with a similar porting, I bumped into this [colab notebook](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/migrate.ipynb#scrollTo=ulaB1ymO4pw5) containing a very juicy paragraph on the subject:

> 

> # A note on Slim & contrib.layers
> 
> A large amount of older TensorFlow 1.x code uses the Slim library, which was packaged with TensorFlow 1.x as tf.contrib.layers. As a contrib module, this is no longer available in TensorFlow 2.0, even in tf.compat.v1. Converting code using Slim to TF 2.0 is more involved than converting repositories that use v1.layers. In fact, it may make sense to convert your Slim code to v1.layers first, then convert to Keras.
> 
>    -  Remove arg_scopes, all args need to be explicit
>    -  If you use them, split normalizer_fn and activation_fn into their own layers
>    -  Separable conv layers map to one or more different Keras layers (depthwise, pointwise, and separable Keras layers)
>    -  Slim and v1.layers have different arg names & default values
>    -  Some args have different scales
>    -  If you use Slim pre-trained models, try out Keras's pre-traimed models from tf.keras.applications or TF Hub's TF2 SavedModels exported from the original Slim code.
> 
> Some tf.contrib layers might not have been moved to core TensorFlow but have instead been moved to the TF add-ons package.

I hope it is useful to other people!

**Comment by draxdave** [comment]
Are you kidding me?? Why this issue is closed while there is no solution yet?

I spent days to learn Python lang to be able to work with TF. Then I followed instructions on installing and running TFv2.0 explain [here](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/installation.md). After facing lots of unexpected issues trying to install TF(illegal instruction and building from source ..) finally I'm done installing it on my Ubuntu server. Now, I have `no attribute 'contrib'`  and interestingly, everyone is suggesting me to downgrade to TFv1.x to pass this stage too. Is't this nasty?
 I'm exhausted trying to launch TF.

**Comment by tomerk** [comment]
Hi Draxdave@, we're sorry you've found this to be a painful experience! Here's some information that might be useful to you or others who come across this issue:

1. That link you put is from the TF object detection library. The team that owns it is actively working on releasing the tf 2.x version of the object detection library + training scripts. If you want to use the training scripts for it that have already been released you need to use TF 1.x.

2. Contrib is not packaged with Tensorflow 2.x. The design discussions around that decision can be found here https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md

3. Due to it's design, TF slim does not work with eager execution enabled. That said, it has been made it available (with extremely light support) as a TF2 compatible library (as long as you call tf.compat.v1.disable_eager_execution()` in https://github.com/google-research/tf-slim

We would welcome community contributions that make tf-slim work with v2 behavior enabled.

**Comment by draxdave** [comment]
@tomerk First of all, thanks for your quick response. 
Yes I'm trying to use Object detection library. This is what I did (may save some time for others) :

- I Installed TFv2.2 as a default released version but it gave me `Illegal instruction` so I built it from the source. After building and installing it, I got `no attribute 'contrib'` Error.
- So I built TFv1.14 release as a tutorial was moving on with and again I had to build it from the source and install it, though I got #32947 issue afterwards. 
- This time I switched on TFv1.15.0rc2 according to an active Tensorflow contributor suggestion [here](https://github.com/tensorflow/tensorflow/issues/32947#issuecomment-538171238) and building from source stuff. I ended up with #30728 Keras issue.
- Now I'm building TFv1.13.1 as some claimed that it works [here](https://github.com/tensorflow/tensorflow/issues/30728#issuecomment-511564565)
- However, I found pre-built TF versions suitable for old processors (cpu/gpu) but they were not compatible with my target servers specs. See #19584 discussion and [here](https://github.com/yaroslavvb/tensorflow-community-wheels/issues/103) 


_My machines specs:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                1
On-line CPU(s) list:   0
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 26
Model name:            Intel Core i7 9xx (Nehalem Class Core i7)
Stepping:              3
CPU MHz:               3292.060
BogoMIPS:              6584.12
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
NUMA node0 CPU(s):     0
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx lm constant_tsc rep_good nopl tsc_known_freq pni ssse3 cx16 sse4_1 sse4_2 x2apic popcnt hypervisor lahf_lm kaiser_


**I'll update this post in case I find a solution other than Pytorch and downgrading**

> Hi Draxdave@, we're sorry you've found this to be a painful experience! Here's some information that might be useful to you or others who come across this issue in the future:
> 
> 1. That link you put is from the TF object detection library. The team that owns it is actively working on releasing the tf 2.x version of the object detection library + training scripts. If you want to use the training scripts for it that have already been released you need to use TF 1.x.
> 2. Contrib is not packaged with Tensorflow 2.x. The design discussions around that decision can be found here https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
> 3. Due to it's design, TF slim does not work with eager execution enabled. That said, it has been made it available (with extremely light support) as a TF 2.0-binary compatible library (as long as you call `tf.compat.v1.disable_eager_execution()`) in https://github.com/google-research/tf-slim
> 
> We would welcome community contributions that make tf-slim work with v2 behavior enabled.

**Comment by manyids2** [comment]
Facing the same problem trying to use Object detection library, has there been any update?

**Comment by draxdave** [comment]
> Facing the same problem trying to use Object detection library, has there been any update?

TFv1.13 seems working for now, however I have not tested it in my real world project yet. Many warnings and rubbish message is being thrown at runtime which you shouldn't care about.

**Comment by Trapti04** [comment]
I am also facing major issues trying to work around getting tf.contrib.slim dependent code to work for my project. I started by downloading tf 1.4 version [ downgrade] but that does not help. I am considering re-writing the model layers in pure tensorflow instead to workaround this issue. Any guidance on how layer functionality from slim packages map to the more tensorflow 2  libraries?

**Comment by brucechou1983** [comment]
I think this is one of the reasons that many new projects (including google research projects) still use TF 1.x. An important issues like this are closed without providing any solution. I believe more and more AI projects will migrate to PyTorch once it has mature deployment tools for edge devices.
